tests/test_otel.py

```diff
@@ -0,0 +1,334 @@
+    "headers", [None, context], ids=["no context", "context"]
+)
+@pytest.mark.parametrize("path", ["/204"], ids=["trace off"])
+def test_no_variables(headers, path, trace_service, client):
+    r = client.get(f"http://127.0.0.1:18080{path}", headers=headers)
+
+    time.sleep(0.5)  # wait for spans
+    assert len(trace_service.spans) == 0
+
+    assert r.headers.get("X-Otel-Trace-Id") is None
+    assert r.headers.get("X-Otel-Span-Id") is None
+    assert r.headers.get("X-Otel-Parent-Id") is None
+    assert r.headers.get("X-Otel-Parent-Sampled") is None
+
+
+@pytest.mark.parametrize(
+    ("headers", "idx"),
+    [(None, 0), (context, 1)],
+    ids=["no context", "context"],
+)
+@pytest.mark.parametrize(
+    ("path", "pid", "tparent", "tstate"),
+    [
+        (
+            "/context-ignore",
+            [None, None],
+            [None, context["Traceparent"]],
+            [None, context["Tracestate"]],
+        ),
+        (
+            "/context-extract",
+            [None, span_id],
+            [None, context["Traceparent"]],
+            [None, context["Tracestate"]],
+        ),
+        (
+            "/context-inject",
+            [None, None],
+            ["00-trace_id-span_id-01", "00-trace_id-span_id-01"],
+            [None, None],
+        ),
+        (
+            "/context-propagate",
+            [None, span_id],
+            ["00-trace_id-span_id-01", f"00-{trace_id}-span_id-01"],
+            [None, context["Tracestate"]],
+        ),
+    ],
+    ids=["ignore", "extract", "inject", "propagate"],
+)
+def test_trace_context(
+    headers, path, idx, pid, tparent, tstate, trace_service, client
+):
+    r = client.get(f"http://127.0.0.1:18080{path}", headers=headers)
+
+    span = trace_service.get_span()
+
+    # Validate headers
+    assert r.headers.get("X-Otel-Parent-Id") == pid[idx]
+    assert r.headers.get("X-Otel-Traceparent") == (
+        "-".join(decode_id(span, val) for val in tparent[idx].split("-"))
+        if tparent[idx] is not None
+        else None
+    )
+    assert r.headers.get("X-Otel-Tracestate") == tstate[idx]
+
+
+@pytest.mark.parametrize(
+    "nginx_config",
+    [
+        ({"interval": "1s"}),
+    ],
+    indirect=True,
+    ids=["batch_size 3"],
+    scope="module",
+)
+@pytest.mark.parametrize(
+    ("path", "n", "nbatch", "nspan"),
+    [
+        ("/", 7, 3, [3] * 2 + [1]),
+        ("/", 16, 6, [3] * 5 + [1]),
+    ],
+    ids=["7 requests", "16 requests"],
+)
+def test_batches(path, n, nbatch, nspan, trace_service, client):
+    for _ in range(n):
+        client.get(f"http://127.0.0.1:18080{path}")
+
+    time.sleep(1.5)  # wait for the rest batches
+    assert len(trace_service.spans) == nbatch
+    for i, batch in enumerate(trace_service.spans):
+        assert (
+            get_attr(batch[0].resource, "service.name", "string_value")
+            == "test_service"
+        )
+        assert len(batch[0].scope_spans[0].spans) == nspan[i]
+    trace_service.spans.clear()
```

Comment:
Assuming some other notes:
```suggestion
@pytest.mark.parametrize("batch_count", [1, 3])
def test_batches(trace_service, client, batch_count):
    batch_size = 3
    
    for _ in range(batch_count * batch_size + 1): # +1 request to trigger batch sending
        assert client.get(f"http://127.0.0.1:18080/ok").status_code == 200
    
    time.sleep(0.01)
    
    for batch in trace_service.batches:
        assert get_attr(batch[0].resource, "service.name") == "test_service"
        assert len(batch[0].scope_spans[0].spans) == batch_size
        
    time.sleep(0.3) # wait for +1 request to be flushed
    trace_service.batches.clear()
```

Reply:
Applied.